File: getting-started/variant-ts/module-09-capstone-wordle/exercise-01-wordle/stub/game/game.ts
import { isValidWord } from "./words";

export const MAX_GUESSES = 6;
export const WORD_LENGTH = 5;

// LetterResult describes how a single letter in a guess relates
// to the target word. A guess produces an array of five of these.
export type LetterResult = "correct" | "misplaced" | "absent";

// GuessResult holds the outcome of evaluating one guess.
export type GuessResult = {
  guess: string;
  letters: LetterResult[];
};

// GameStatus represents the three possible states of a game.
// A game that's still going has guesses remaining and no win.
export type GameStatus = "playing" | "won" | "lost";

// Game holds the complete state of a Wordle game.
// The target word is hidden -- callers interact through makeGuess.
export class Game {
  #target: string;
  #guesses: GuessResult[] = [];
  #status: GameStatus = "playing";

  // The target must be exactly WORD_LENGTH letters.
  constructor(target: string) {
    // TODO: validate target length, store lowercase.
    this.#target = target;
  }

  // makeGuess processes a guess and returns the result.
  // Returns an error string if:
  //   - the game is already over (won or lost)
  //   - the guess is not exactly WORD_LENGTH letters
  //   - the guess is not a valid word (use isValidWord)
  makeGuess(guess: string): { result: GuessResult } | { error: string } {
    // TODO: validate the guess, evaluate it, update game state.
    void isValidWord;
    return { error: "not implemented" };
  }

  get status(): GameStatus { return this.#status; }
  get guesses(): GuessResult[] { return this.#guesses; }
  get attemptsRemaining(): number { return MAX_GUESSES - this.#guesses.length; }
}

// evaluateGuess compares a guess against a target and returns
// the letter-by-letter result. This is the core algorithm.
//
// Rules:
//   - A letter in the correct position is "correct".
//   - A letter that exists in the target but in a different position
//     is "misplaced" -- BUT only if that letter hasn't already been
//     matched by a "correct" or a previous "misplaced".
//   - All other letters are "absent".
//
// Examples:
//   target: "apple"
//   guess:  "paper"
//   result: ["misplaced", "misplaced", "correct", "misplaced", "absent"]
//
//   p -> misplaced (exists in "apple" at index 2, but guess has it at 0)
//   a -> misplaced (exists in "apple" at index 0, but guess has it at 1)
//   p -> correct   (index 2 matches)
//   e -> misplaced (exists in "apple" at index 4, guess has it at 3)
//   r -> absent    (not in "apple")
//
//   target: "hello"
//   guess:  "llama"
//   result: ["misplaced", "misplaced", "absent", "absent", "absent"]
//
//   First 'l' -> misplaced (consumes the 'l' at target index 2)
//   Second 'l' -> misplaced (consumes the 'l' at target index 3)
//   'a' -> absent
//   'm' -> absent
//   'a' -> absent
//
//   "hello" has two l's, so both l's get marked.
//   If the target only had one 'l', only the first would be misplaced.
//   Think carefully about duplicate letters.
export const evaluateGuess = (target: string, guess: string): LetterResult[] => {
  // TODO: implement the evaluation algorithm.
  // Hint: do two passes. First pass: mark all correct letters.
  // Second pass: for each non-correct letter, check if it's
  // misplaced (exists in target and hasn't been "consumed" yet).
  return Array.from({ length: WORD_LENGTH }, () => "absent" as LetterResult);
};

File: getting-started/variant-go/module-09-capstone-wordle/exercise-01-wordle/stub/game/game.go
package game

const MaxGuesses = 6
const WordLength = 5

// LetterResult describes how a single letter in a guess relates
// to the target word. A guess produces a slice of five of these.
type LetterResult int

const (
	Absent    LetterResult = iota // letter is not in the target word
	Misplaced                     // letter is in the word but wrong position
	Correct                       // letter is in the correct position
)

func (r LetterResult) String() string {
	return [...]string{"absent", "misplaced", "correct"}[r]
}

// GuessResult holds the outcome of evaluating one guess.
type GuessResult struct {
	Guess   string
	Letters [WordLength]LetterResult
}

// GameStatus represents the three possible states of a game.
// A game that's still going has guesses remaining and no win.
// This is a sealed set — only these three values exist.
type GameStatus int

const (
	Playing GameStatus = iota
	Won
	Lost
)

// Game holds the complete state of a Wordle game.
// The target word is hidden — callers interact through MakeGuess.
type Game struct {
	target  string
	guesses []GuessResult
	status  GameStatus
}

// New creates a game with the given target word.
// The target must be exactly WordLength letters.
func New(target string) *Game {
	// TODO: create and return a new game.
	// Validate that target is the right length.
	return nil
}

// MakeGuess processes a guess and returns the result.
// Returns an error if:
//   - the game is already over (Won or Lost)
//   - the guess is not exactly WordLength letters
//   - the guess is not a valid word (use IsValidWord)
func (g *Game) MakeGuess(guess string) (GuessResult, error) {
	// TODO: validate the guess, evaluate it, update game state.
	return GuessResult{}, nil
}

// EvaluateGuess compares a guess against a target and returns
// the letter-by-letter result. This is the core algorithm.
//
// Rules:
//   - A letter in the correct position is Correct.
//   - A letter that exists in the target but in a different position
//     is Misplaced — BUT only if that letter hasn't already been
//     matched by a Correct or a previous Misplaced.
//   - All other letters are Absent.
//
// Example:
//   target: "apple"
//   guess:  "paper"
//   result: [Misplaced, Misplaced, Correct, Absent, Absent]
//
//   The first 'p' is Misplaced (exists in target, wrong position).
//   The 'a' is Misplaced (exists in target, wrong position).
//   The second 'p' is Correct (right position).
//   The 'e' is Absent — even though 'e' is in "apple", it's at
//   position 4 and the guess has it at position 3. But wait, there's
//   only one 'e' in the target word, and... this is the tricky part.
//   Think carefully about duplicate letters.
func EvaluateGuess(target, guess string) [WordLength]LetterResult {
	// TODO: implement the evaluation algorithm.
	// Hint: do two passes. First pass: mark all Correct letters.
	// Second pass: for each non-Correct letter, check if it's
	// Misplaced (exists in target and hasn't been "consumed" yet).
	var result [WordLength]LetterResult
	return result
}

// Status returns the current game status.
func (g *Game) Status() GameStatus { return g.status }

// Guesses returns all guesses made so far.
func (g *Game) Guesses() []GuessResult { return g.guesses }

// AttemptsRemaining returns how many guesses are left.
func (g *Game) AttemptsRemaining() int {
	return MaxGuesses - len(g.guesses)
}
